### PR TITLE
Fix reading of black-and-white (thresholded) TIFF images

### PR DESCRIPTION
**Pull request by venalil**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/538

I recently updated my local OpenCV version to 3.4.3 and found out that
I could not read my TIFF images related to my project. After debugging I
found out that there has been some static analysis fixes made
that accidentally have broken reading those black-and-white TIFF images.

Commit hash in which reading of mentioned TIFF images has been broken:
cbb1e867e5141412c62ff534def7f117e28e04e8

Basically the fix is to revert back to the same functionality that has been there before.
When black-and-white images are read bpp (bitspersample) is 1.
Without the case 1: this TiffDecoder::readHeader() function always return false.

### This pullrequest changes

PR fixes reading thresholded (black-and-white) TIFF images.

```
opencv_extra=3.4
```

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 1]
```diff
@@ -253,6 +253,9 @@ bool TiffDecoder::readHeader()
             int wanted_channels = normalizeChannelsNumber(ncn);
             switch(bpp)
             {
+                case 1:
+                    result = true;
+                    break;
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp]: What is about "m_type" ?
It must be initialized in this method.

Also it would be nice to add proper error reporting via "default:" case.

**Comment by alalek** [comment]
Could you please provide some sample image with mentioned problem (64x64 pixels is enough) ? It would be nice to use it in tests.

**Comment by alalek** [comment]
This patch should [go](https://github.com/opencv/opencv/wiki/How_to_contribute#faq) into 3.4 branch first. We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/imgcodecs/src/grfmt_tiff.cpp (hunk 1 above)
Comment (venalil) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: It was not previously initialized so tbh I have not figured it out.

Yes I agree the default case should have some error message because now it silently failed and it was frustrating.

I will figure these things out when I can.

Comment (alalek) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: Actually this case was not properly supported (bpp=1 case works "by mistake").
Also there are no tests for this type of images (btw, imgcodecs test data is [here](https://github.com/opencv/opencv_extra/tree/3.4/testdata/highgui/readwrite)).

Comment (venalil) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: Yup I know it worked by mistake, just wanted to revert the changes that broke this functionality which the original committer was unwilling to do :)

So this will take some time on my behalf...

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 2]
```diff
@@ -269,6 +273,8 @@ bool TiffDecoder::readHeader()
                     m_type = CV_MAKETYPE(CV_64F, photometric > 1 ? 3 : 1);
                     result = true;
                     break;
+            default:
+                throw std::runtime_error("Invalid bitsperpixel value read from TIFF header! Must be 1, 8, 16, 32 or 64.");
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp]: Use `CV_Error()` for error reporting.

Comment (venalil) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: Test data: https://github.com/opencv/opencv_extra/pull/538

Ok I'll change that

**Comment by alalek** [comment]
@venalil Well done!

I added test for image from opencv_extra and pushed into this PR.
